app/src/main/java/agency/five/codebase/android/movieapp/ui/home/HomeViewModel.kt

```diff
@@ -0,0 +1,128 @@
+
+    private val _upcomingMovieHomeViewState = MutableStateFlow(HomeMovieCategoryViewState(emptyList(), emptyList()))
+    val upcomingMovieHomeViewState: StateFlow<HomeMovieCategoryViewState> =
+        _upcomingMovieHomeViewState.asStateFlow()
+
+    init {
+        initializeMovieCategories()
+    }
+
+    private fun initializeMovieCategories() {
+        getPopular(homeScreenMapper)
+        getUpcoming(homeScreenMapper)
+        getNowPlaying(homeScreenMapper)
+    }
+
+    private fun getPopular(homeScreenMapper: HomeScreenMapper) {
+        viewModelScope.launch {
+            movieRepository.popularMovies(MovieCategory.POPULAR_STREAMING).collect {
+                _popularMoviesHomeViewState.value = homeScreenMapper.toHomeMovieCategoryViewState(
+                    movieCategories = listOf(
+                        MovieCategory.POPULAR_STREAMING,
+                        MovieCategory.POPULAR_ON_TV,
+                        MovieCategory.POPULAR_FOR_RENT,
+                        MovieCategory.POPULAR_IN_THEATERS
+                    ),
+                    selectedMovieCategory = popularSelectedCategory.value,
+                    movies = it
+                )
+            }
+        }
+    }
+
+    private fun getNowPlaying(homeScreenMapper: HomeScreenMapper) {
+        viewModelScope.launch {
+            movieRepository.nowPlayingMovies(MovieCategory.NOW_PLAYING_MOVIES).collect {
+                _nowPlayingMoviesHomeViewState.value =
+                    homeScreenMapper.toHomeMovieCategoryViewState(
+                        movieCategories = listOf(
+                            MovieCategory.NOW_PLAYING_MOVIES,
+                            MovieCategory.NOW_PLAYING_TV,
+                        ),
+                        selectedMovieCategory = nowPlayingSelectedCategory.value,
+                        movies = it
+                    )
+            }
+        }
+    }
+
+    private fun getUpcoming(homeScreenMapper: HomeScreenMapper) {
+        viewModelScope.launch {
+            movieRepository.upcomingMovies(MovieCategory.UPCOMING_TODAY).collect {
+                _upcomingMovieHomeViewState.value = homeScreenMapper.toHomeMovieCategoryViewState(
+                    movieCategories = listOf(
+                        MovieCategory.UPCOMING_TODAY,
+                        MovieCategory.UPCOMING_THIS_WEEK,
+                    ),
+                    selectedMovieCategory = upcomingSelectedCategory.value,
+                    movies = it
+                )
+            }
+        }
+    }
+
+    fun changeCategory(categoryId: Int){
+        when (categoryId) {
+            MovieCategory.POPULAR_STREAMING.ordinal,
+            MovieCategory.POPULAR_FOR_RENT.ordinal,
+            MovieCategory.POPULAR_ON_TV.ordinal,
+            MovieCategory.POPULAR_IN_THEATERS.ordinal
+            -> {
+                popularSelectedCategory.value = MovieCategory.values()[categoryId]
+                getPopular(homeScreenMapper)
```

Comment:
You shouldn't call `getPopular()` method here. Instead, you should emit newly selected movie category to `popularSelectedCategory`, and this should cause updating `popularMoviesHomeViewState`. In order to achieve that, you need to define `popularSelectedCategory` as `MutableStateFlow`, and than use this flow together with `movieRepository.popularMovies(...)` flow in order to produce `popularMoviesHomeViewState` . You can achieve this with `flatMapLatest` and `stateIn` operators like this:

```kotlin
private val popularMoviesCategorySelected = MutableStateFlow(MovieCategory.POPULAR_STREAMING)

val popularCategoryViewState: StateFlow<HomeMovieCategoryViewState> =
    popularMoviesCategorySelected
        .flatMapLatest { selectedMovieCategory ->
            movieRepository.popularMovies(selectedMovieCategory)
                .map { popularMovies ->
                    homeScreenMapper.toHomeMovieCategoryViewState(
                        movieCategories = popularMovieCategories, // listOf(MovieCategory.POPULAR ...)
                        selectedMovieCategory = selectedMovieCategory,
                        movies = popularMovies,
                    )
                }
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5_000),
            initialValue = homeScreenMapper.toHomeMovieCategoryViewState(
                movieCategories = popularMovieCategories, // listOf(MovieCategory.POPULAR ...)
                selectedMovieCategory = popularMoviesCategorySelected.value,
                movies = emptyList(),
            )
        )
```

With this approach, you won't need to have `getPopular(...)` and `initializeMovieCategories()` functions (as well as `getUpcoming` and `getNowPlaying`).

Reply:
You should define this behaviour for producing view states to all your view models (`HomeViewModel`, `FavoritesViewModel`, `MovieDetailsViewModel`).